python/ray/data/aggregate.py

```diff
@@ -71,17 +79,168 @@ def __init__(self):
+    1. If r is null and ignore_nulls=False, return None.
+    2. If r is null and ignore_nulls=True, return a.
+    3. If r is non-null and a is None, return None.
+    5. If r is non-null and a is non-None, return accum(a[:-1], r).
+
+    Args:
+        ignore_nulls: Whether nulls should be ignored or cause a null result.
+        on_fn: Function selecting a subset of the row to apply the aggregation.
+        accum: The core accumulator function.
+        a: Accumulated so far.
+        r: This row.
+
+    Returns:
+        Accumulated result including the provided row.
+    """
+    r = on_fn(r)
+    if _is_null(r):
+        if ignore_nulls:
+            return a
+        else:
+            return None
+    else:
+        if a is None:
+            return None
+        else:
+            a = a[:-1]
+            if len(a) == 1:
+                a = a[0]
+            res = accum(a, r)
+            if not isinstance(res, list):
+                res = [res]
+            return res + [1]
+
+
+def _null_merge(
+    ignore_nulls: bool,
+    merge: Callable[[AggType, AggType], AggType],
+    a1: AggType,
+    a2: AggType,
+) -> AggType:
+    """
+    Merge two accumulations while handling nulls.
+
+    Expects a1 and a2 to be either None or of the form:
+    a = [acc_data_1, ..., acc_data_2, has_data].
+
+    This merges two accumulations subject to the following null rules:
+    1. If a1 is empty and a2 is empty, return empty accumulation.
+    2. If a1 (a2) is empty and a2 (a1) is None, return None.
+    3. If a1 (a2) is empty and a2 (a1) is non-None, return a2 (a1).
+    4. If a1 (a2) is None, return a2 (a1) if ignoring nulls, None otherwise.
+    5. If a1 and a2 are both non-null, return merge(a1, a2).
+
+    Args:
+        ignore_nulls: Whether nulls should be ignored or cause a None result.
+        merge: The core merge function.
+        a1: One of the intermediate accumulations to merge.
+        a2: One of the intermediate accumulations to merge.
+
+    Returns:
+        Accumulation of the two provided accumulations.
+    """
+    if a1 is None:
+        # If we're ignoring nulls, propagate a2; otherwise, propagate None.
+        return a2 if ignore_nulls else None
+    if a1[-1] == 0:
+        # If a1 is empty, propagate a2.
+        # No matter whether a2 is a real value, the empty, or None,
+        # propagating each of these is correct if a1 is empty.
+        return a2
+    if a2 is None:
+        # If we're ignoring nulls, propagate a1; otherwise, propagate None.
+        return a1 if ignore_nulls else None
+    if a2[-1] == 0:
+        # If a2 is empty, propagate a1.
+        return a1
+    a1 = a1[:-1]
+    if len(a1) == 1:
+        a1 = a1[0]
+    a2 = a2[:-1]
+    if len(a2) == 1:
+        a2 = a2[0]
+    res = merge(a1, a2)
+    if not isinstance(res, list):
+        res = [res]
+    return res + [1]
+
+
+def _null_finalize(finalize: Callable[[AggType], AggType], a: AggType) -> AggType:
```

Comment:
finalize returns `U` instead of `AggType`

Reply:
Great catch!